controlplane/kubeadm/internal/controllers/remediation.go

```diff
@@ -191,9 +211,107 @@ func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.C
+func (r *KubeadmControlPlaneReconciler) checkRetryLimits(log logr.Logger, machineToBeRemediated *clusterv1.Machine, controlPlane *internal.ControlPlane) (hasRetries bool, retryCount int32) {
+	// If there is no last remediation, this is the first try of a new retry sequence.
+	if controlPlane.KCP.Status.LastRemediation == nil {
+		return true, 0
+	}
+
+	// Gets MinHealthySeconds and RetryDelaySeconds from the remediation strategy, or use defaults.
+	minHealthyPeriod := controlplanev1.DefaultMinHealthyPeriod
+	if controlPlane.KCP.Spec.RemediationStrategy != nil && controlPlane.KCP.Spec.RemediationStrategy.MinHealthyPeriod != nil {
+		minHealthyPeriod = controlPlane.KCP.Spec.RemediationStrategy.MinHealthyPeriod.Duration
+	}
+
+	retryDelay := time.Duration(0)
+	if controlPlane.KCP.Spec.RemediationStrategy != nil {
+		retryDelay = controlPlane.KCP.Spec.RemediationStrategy.RetryPeriod.Duration
+	}
+
+	// Gets the timestamp of the last remediation; if missing, default to a value
+	// that ensures both MinHealthySeconds and RetryDelaySeconds are expired.
+	// NOTE: this could potentially lead to executing more retries than expected or to executing retries before than
+	// expected, but this is considered acceptable.
+	// when the system recovers from someone/something manually removing the LastRemediatingTimeStampAnnotation.
+	max := func(x, y time.Duration) time.Duration {
+		if x < y {
+			return y
+		}
+		return x
+	}
+
+	lastRemediationTimestamp := time.Now().Add(-2 * max(minHealthyPeriod, retryDelay)).UTC()
+	if controlPlane.KCP.Status.LastRemediation != nil {
+		lastRemediationTimestamp = time.Unix(controlPlane.KCP.Status.LastRemediation.Timestamp.Seconds, int64(controlPlane.KCP.Status.LastRemediation.Timestamp.Nanos))
+	}
+
+	// Check if the machine being remediated has been created as a remediation for a previous unhealthy machine.
+	// NOTE: if someone/something manually removing the RemediationForAnnotation on Machines or one of the LastRemediatedMachineAnnotation
+	// and LastRemediatedMachineRetryAnnotation on KCP, this could potentially lead to executing more retries than expected,
+	// but this is considered acceptable in such a case.
+	machineRemediatingFor := machineToBeRemediated.Name
+	if remediationFor, ok := machineToBeRemediated.Annotations[controlplanev1.RemediationForAnnotation]; ok {
+		// If the remediation is happening before minHealthyPeriod is expired, then KCP considers this
+		// as a remediation for the same previously unhealthy machine.
+		// TODO: add example
+		if lastRemediationTimestamp.Add(minHealthyPeriod).After(time.Now()) {
+			machineRemediatingFor = remediationFor
+			log.WithValues("RemediationRetryFor", klog.KRef(machineToBeRemediated.Namespace, machineRemediatingFor))
+		}
+	}
+
+	// If remediation is happening for a different machine, this is the first try of a new retry sequence.
+	if controlPlane.KCP.Status.LastRemediation.Machine != machineRemediatingFor {
```

Comment:
I think this doesn't work correctly if two different machines are failing over time

Example:
* M1, M2, M3 are healthy
* M1 is reconciled to M1-1 (KCP.status.lastRemediation.machine = M1)
* M2 is reconciled to M2-1 (KCP.status.lastRemediation.machine = M2)
* M1-1 is reconciled to M1-2
  * Because KCP.status.lastRemediation.machine is pointing to M2, we don't handle the retry of the M1-1 correctly (we don't detect that it's another retry)

I think we have to decide:
* Either we count retries only once for the entire KCP (i.e. we only have one counter for all of KCP)
* Or we count retries per Machine (but than we can't just have one LastRemedation field in the KCP.status as it's not able to track retries of different machines)

Reply:
Currently, we are in the first use case (we count retries only once for the entire KCP), and this should be enough given that KCP mostly acts always changing one machine at a time, and then waiting for everything to become stable before doing the next actions.
We can eventually reconsider in the future to have more counters if we got evidence that remediation flickering from one machine to another is a common use case, but I will rule it out for this first iteration.

Reply:
Ah so it's intentional that we count retries etcd just for a specific machine and if another one starts failing we reset the counter?

So we're kind of counting for the entire KCP, but as soon as another Machine fails we start from 0.

Reply:
This got fixed while moving last remediation data to machine level